Refactor self-alignment property resolution to manage 'auto' instead of 'normal'

fantasai opened the pull request:
#### fd2edbac21f5950d58d0720afc0471640061a6e7
<pre>
Refactor self-alignment property resolution to manage &apos;auto&apos; instead of &apos;normal&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=304716">https://bugs.webkit.org/show_bug.cgi?id=304716</a>
<a href="https://rdar.apple.com/167221674">rdar://167221674</a>

Reviewed by Sammy Gill.

The way &apos;normal&apos; resolution is looped through the style system ties into our
layout code in an awkward way. Instead, loop &apos;auto&apos; resolution through the
style system, and make layout code handle &apos;normal&apos; (which is much more
tied to formatting context subtleties) directly. This creates more
straightforward logic in both systems, divides the work along a more
appropriate boundary, and also avoids doing &apos;normal&apos; resolution work when
we don&apos;t need it.

This is Part 1 of a refactoring stack for grid-lanes item sizing (bug 304715).

* Source/WebCore/layout/formattingContexts/flex/FlexLayout.cpp:
(WebCore::Layout::FlexLayout::crossSizeForFlexLines const):
(WebCore::Layout::FlexLayout::computeCrossSizeForFlexItems const):
(WebCore::Layout::FlexLayout::handleCrossAxisAlignmentForFlexItems const):
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::remainingSpaceForStaticAlignment const):
(WebCore::PositionedLayoutConstraints::resolveAlignmentValue const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::overflowAlignmentForFlexItem const):
(WebCore::RenderFlexibleBox::alignmentForFlexItem const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::styleDidChange):
(WebCore::RenderGrid::columnAxisPositionForGridItem const):
(WebCore::RenderGrid::alignSelfForGridItem const): Deleted.
(WebCore::RenderGrid::justifySelfForGridItem const): Deleted.
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
(WebCore::StyleSelfAlignmentData::isNormal const):
(WebCore::StyleSelfAlignmentData::isStretch const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/values/align/StyleAlignItems.cpp:
(WebCore::Style::AlignItems::resolve const):
* Source/WebCore/style/values/align/StyleAlignItems.h:
* Source/WebCore/style/values/align/StyleAlignSelf.cpp:
(WebCore::Style::AlignSelf::resolve const):
* Source/WebCore/style/values/align/StyleAlignSelf.h:
* Source/WebCore/style/values/align/StyleJustifyItems.cpp:
(WebCore::Style::JustifyItems::resolve const):
* Source/WebCore/style/values/align/StyleJustifyItems.h:
* Source/WebCore/style/values/align/StyleJustifySelf.cpp:
(WebCore::Style::JustifySelf::resolve const):
* Source/WebCore/style/values/align/StyleJustifySelf.h:

Canonical link: <a href="https://commits.webkit.org/305146@main">https://commits.webkit.org/305146@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4874a64e84b74bc3d523a16099eaf21b32efaa74

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137625 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/145383 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90598 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2c9d573d-c4b6-4d03-b826-68fe3b1c9ad9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10116 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105261 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f3d018a7-9cc0-4545-8b4c-a91c6276b120) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140570 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86116 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9cf789d1-c3b8-4105-b7a9-0ffb93836efe) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7571 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5294 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5966 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148150 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9669 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42051 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113642 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9686 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113985 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28936 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7501 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119580 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64333 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9717 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37629 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9448 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73282 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9657 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9509 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
```diff
@@ -60,13 +60,18 @@ class StyleSelfAlignmentData {
+            return normal == ItemPosition::Stretch;
+        return position() == ItemPosition::Stretch;
+    }
+
```
Hmmm I feel like this new API shares some characteristics with the old `isNormal` API that we're modifying... 

At the call sites it could be a bit confusing when you have something like `if (alignSelf.isStretch(ItemPosition::Stretch)` it's not immediately clear what the argument is doing in relation to checking if align-self is stretch. If we force the callers to handle normal as stretch explicitly whenever required by the formatting context maybe that would keep things clear. WDYT?

fantasai replied:
I see what you're saying... The flip side is that it makes things with more complex normal behavior more convoluted to write on the calling side. Consider this for example:

```
    if (!itemStyle.alignSelf().resolve(&style()).isStretch(mode == StretchingMode::Explicit ? ItemPosition::Normal : ItemPosition::Stretch))
       ... do something ...
```

If we didn't have the ability to pass in the normal behavior, the caller would have to write something like

```
   auto alignment = itemStyle.alignSelf().resolve(&style());
   auto normal = mode == StretchingMode::Explicit ? ItemPosition::Normal : ItemPosition::Stretch; // or maybe this is a function call to more complicated logic
   if (!(alignment.isNormal() && normal = ItemPosition::Stretch) && !alignment.isStretch())
      ... do something ...
```

I could go either way, but that's why I wrapped the normal check into this helper: it makes it easier to write straightforward conditionals. (By, as you note, hiding a bit of the logic.)

fantasai replied:
Though now that I'm thinking about it, even if we keep it this way, maybe I can rename it to `isStretching()` so it's a bit more obvious that it's not necessarily just checking for ItemPosition::Stretch...

sammygill replied:
Oh yeah I see! We probably don't need to make it perfect in this PR and could revisit it if we wanted to, but a rename would probably be nice so that it indicates there might be a bit more work being done here than just checking the `ItemPosition`

fantasai replied:
Ok, added both `isStretch()` for  a simple check against `ItemPosition::Stretch` and `isStretchy(ItemPosition)` to incorporate the normal check!

sammygill commented on Source/WebCore/style/values/align/StyleAlignSelf.cpp:
```diff
@@ -50,9 +54,9 @@ StyleSelfAlignmentData AlignSelf::resolve(std::optional<StyleSelfAlignmentData>
     switch (primary()) {
     case PrimaryKind::Auto:
-        return valueForNormalOrAuto.value_or(ItemPosition::Auto);
+        return { ItemPosition::Auto };
```
Maybe we could add an `ASSERT_NOT_REACHED` here since it doesn't seem like we can reach this code anymore. Would force us to double check any change in the future that would end up falling into this case.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -154,7 +154,13 @@ void RenderGrid::styleDidChange(Style::Difference diff, const RenderStyle* oldSt
+    auto oldJustifyItems = oldStyle->justifyItems().resolve().position();
+    auto newJustifyItems = style().justifyItems().resolve().position();
+    bool justifyStretchChanged = oldJustifyItems != newJustifyItems && (oldJustifyItems == ItemPosition::Normal || newJustifyItems == ItemPosition::Normal || oldJustifyItems == ItemPosition::Stretch || newJustifyItems == ItemPosition::Stretch);
+    if (alignStretchChanged || justifyStretchChanged) {
```
Technically you could put the `alignStretchChanged` and `justifyStretchChanged` each into lambdas and call them in the if statement. That way if the first returns true we don't run the second but it's not likely it's very expensive anyways so up to you

sammygill replied:
Also maybe the variables could make it clear that we're checking `align-items` and `justify-items` when looking to see if it's a stretch change

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -374,29 +374,16 @@ void RenderFlexibleBox::styleDidChange(Style::Difference diff, const RenderStyle
-    auto oldStyleAlignItemsIsStretch = oldStyle->alignItems().resolve(selfAlignmentNormalBehavior()).position() == ItemPosition::Stretch;
+    auto oldAlignment = oldStyle->alignItems().resolve().position();
+    auto newAlignment = style().alignItems().resolve().position();
+    auto stretchAlignmentChanged = (oldAlignment == ItemPosition::Normal || oldAlignment == ItemPosition::Stretch) != (newAlignment == ItemPosition::Normal || newAlignment == ItemPosition::Stretch);
```
Maybe the variable names could be a bit more explicit that it's alignment from `align-items`

fantasai committed:
Refactor self-alignment property resolution to manage 'auto' instead of 'normal'
https://bugs.webkit.org/show_bug.cgi?id=304716
rdar://167221674

Reviewed by Sammy Gill.

The way 'normal' resolution is looped through the style system ties into our
layout code in an awkward way. Instead, loop 'auto' resolution through the
style system, and make layout code handle 'normal' (which is much more
tied to formatting context subtleties) directly. This creates more
straightforward logic in both systems, divides the work along a more
appropriate boundary, and also avoids doing 'normal' resolution work when
we don't need it.

This is Part 1 of a refactoring stack for grid-lanes item sizing (bug 304715).

* Source/WebCore/layout/formattingContexts/flex/FlexLayout.cpp:
(WebCore::Layout::FlexLayout::crossSizeForFlexLines const):
(WebCore::Layout::FlexLayout::computeCrossSizeForFlexItems const):
(WebCore::Layout::FlexLayout::handleCrossAxisAlignmentForFlexItems const):
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::remainingSpaceForStaticAlignment const):
(WebCore::PositionedLayoutConstraints::resolveAlignmentValue const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::overflowAlignmentForFlexItem const):
(WebCore::RenderFlexibleBox::alignmentForFlexItem const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::styleDidChange):
(WebCore::RenderGrid::columnAxisPositionForGridItem const):
(WebCore::RenderGrid::alignSelfForGridItem const): Deleted.
(WebCore::RenderGrid::justifySelfForGridItem const): Deleted.
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
(WebCore::StyleSelfAlignmentData::isNormal const):
(WebCore::StyleSelfAlignmentData::isStretch const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/values/align/StyleAlignItems.cpp:
(WebCore::Style::AlignItems::resolve const):
* Source/WebCore/style/values/align/StyleAlignItems.h:
* Source/WebCore/style/values/align/StyleAlignSelf.cpp:
(WebCore::Style::AlignSelf::resolve const):
* Source/WebCore/style/values/align/StyleAlignSelf.h:
* Source/WebCore/style/values/align/StyleJustifyItems.cpp:
(WebCore::Style::JustifyItems::resolve const):
* Source/WebCore/style/values/align/StyleJustifyItems.h:
* Source/WebCore/style/values/align/StyleJustifySelf.cpp:
(WebCore::Style::JustifySelf::resolve const):
* Source/WebCore/style/values/align/StyleJustifySelf.h:

Canonical link: https://commits.webkit.org/305146@main
